pkg/egressgateway/manager.go

```diff
@@ -34,94 +20,96 @@ var (
 	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egressgateway")
 )
 
-// The egressgateway manager stores the internal data tracking the policy
-// and endpoint mappings. It also hooks up all the callbacks to update
-// egress bpf map accordingly.
+type k8sCacheSyncedChecker interface {
+	K8sCacheIsSynced() bool
+}
+
+// The egressgateway manager stores the internal data tracking policies and
+// endpoints. It also hooks up all the callbacks to update egress bpf policy map
+// accordingly.
 type Manager struct {
 	mutex lock.Mutex
 
-	// Stores endpoint to policy mapping
-	policyEndpoints map[endpointID][]policyID
-	// Stores policy configs indexed by policyID
+	// k8sCacheSyncedChecker is used to check if the agent has synced its
+	// cache with the k8s API server
+	k8sCacheSyncedChecker k8sCacheSyncedChecker
+
+	// policyConfigs stores policy configs indexed by policyID
 	policyConfigs map[policyID]*PolicyConfig
-	// Stores endpointId to endpoint metadata mapping
+
+	// epDataStore stores endpointId to endpoint metadata mapping
 	epDataStore map[endpointID]*endpointMetadata
 }
 
-func NewEgressGatewayManager() *Manager {
-	return &Manager{
-		policyEndpoints: make(map[endpointID][]policyID),
-		policyConfigs:   make(map[policyID]*PolicyConfig),
-		epDataStore:     make(map[endpointID]*endpointMetadata),
+// NewEgressGatewayManager returns a new Egress Gateway Manager.
+func NewEgressGatewayManager(k8sCacheSyncedChecker k8sCacheSyncedChecker) *Manager {
+	manager := &Manager{
+		k8sCacheSyncedChecker: k8sCacheSyncedChecker,
+		policyConfigs:         make(map[policyID]*PolicyConfig),
+		epDataStore:           make(map[endpointID]*endpointMetadata),
 	}
+
+	manager.runReconciliationAfterK8sSync()
+
+	return manager
+}
+
+// runReconciliationAfterK8sSync spawns a goroutine that waits for the agent to
+// sync with k8s and then runs the first reconciliation.
+func (manager *Manager) runReconciliationAfterK8sSync() {
+	go func() {
+		for range time.NewTicker(time.Second * 1).C {
```

Comment:
You'll need to `Stop()` the ticker, otherwise (I think) this leaks resources.You could just do loop with time.Sleep instead...

Though in this case, since `K8sCacheIsSynced` is just waiting on `d.k8sCachesSynced`, 
why not just do a blocking receive on `d.k8sCachesSynced` instead?

E.g. add something like:
```
func (d *Daemon) WaitUntilK8sCacheIsSynced() {
  _, _ = <-d.k8sCachesSynced
}
```

Reply:
Good shout, switched to `WaitUntilK8sCacheIsSynced`  :+1: